PARQUET-686: Add Order to store the order used for min/max stats.

rdblue opened the pull request:
This adds a new enum, `Order`, that will be set to the order used to produce the min and max values in all `Statistics` objects (at the page level). `Order` has 8 symbols: `SIGNED`, `UNSIGNED`, and 6 symbols for custom orderings. This also adds a `CustomOrder` struct that is used to map the custom order symbols to string descriptors, such as [order keywords used by ICU collating sequences](http://userguide.icu-project.org/collation/api#TOC-Instantiating-the-Predefined-Collators). `CustomOrder` mappings are stored in the file footer.

rdblue commented:
@isnotinvain and @julienledem, could you have a look?

julienledem commented on src/main/thrift/parquet.thrift:
```diff
@@ -197,18 +197,78 @@ enum FieldRepetitionType {
+  CUSTOM3 = 4;
+  CUSTOM4 = 5;
+  CUSTOM5 = 6;
+  CUSTOM6 = 7;
```
I'm not a big fan of this mechanism for extending ordering.
How about we define Collation as a string defined by the unicode standard as in your link bellow:
See http://userguide.icu-project.org/collation/api#TOC-Instantiating-the-Predefined-Collators
And we define "BINARY" as the unsigned order and "SIGNED_BINARY" as the signed one.

julienledem replied:
or possibly have more of a union style struct:
```
enum CollationStandard {
  BINARY, // only "SIGNED" and "UNSIGNED" are accepted
  UNICODE; // must provide locale
}
struct Collation {
  1: required CollationStandard std;
  2: required string name;
}
```

piyushnarang replied:
so if I understand right, we can't have more than 6 custom orders for a given parquet file right?

majetideepak replied:
I like this union style struct approach as well.

rdblue replied:
The problem this enum is trying to solve is to avoid writing collation identifier strings in the metadata of each page. The approach I took was to keep that field small with an enum and translate the custom symbols to collating sequence identifier strings using metadata in the footer.

If we used the union-style struct approach, then "signed", "unsigned", or the identifier string would be present in every page. At a minimum, I think signed and unsigned should be symbols and only custom strings would be in every page.

julienledem replied:
I don't think the sort order should be allowed to change per page. So it should not be in the page metadata but in the footer per column. For example in ColumnMetaData or in SchemaElement.
I would tend to pick SchemaElement.
As a separate note, if it is an id to lookup in a table then we should make it an integer instead of an enum.

piyushnarang commented on src/main/thrift/parquet.thrift:
```diff
@@ -197,18 +197,78 @@ enum FieldRepetitionType {
+   * If this field is missing, the order must be the signed ordering for
+   * backward-compatibility.
+   */
+  5: optional Order order;
```
wondering if there would be a scenario where users want to store multiple ordering stats? (e.g. to support multiple kinds of queries?)

rdblue replied:
I don't think we will need them. Can you think of a specific use case for it?

piyushnarang replied:
Yeah I don't have a concrete use-case for it right now but I think it might be possible that users want to support multiple kinds of queries over their data (one with locale = x, one with locale = y)? Don't think we need to handle this right now though.

piyushnarang commented on src/main/thrift/parquet.thrift:
```diff
@@ -197,18 +197,78 @@ enum FieldRepetitionType {
+
+  /**
+   * A string that identifies the order denoted by the place-holder order enum
+   * symbol. Ordering descriptors should be well-known identifiers, such as the
```
should we be explicit about exactly what the identifiers should be? The comment as it stands now seems open to interpretation. If we explicitly say the identifiers should be only the locale keywords in ICU-54, that allows various language implementations to code on that.

rdblue commented:
Sorry for the delay, I didn't get a notification that there were comments. I'll look through and respond. Thanks for reviewing, everyone!

rdblue commented:
@julienledem, @piyushnarang, can you both take another look at this? I'd like to get this fixed so we can release it.

lekv commented on src/main/thrift/parquet.thrift:
```diff
@@ -197,18 +197,78 @@ enum FieldRepetitionType {
   REPEATED = 2;
 }
 
+/** Identifier for the sort order used to produce min and max values. */
```
Wouldn't this also apply to the sorting_columns_ in the row groups?

lekv replied:
Wouldn't this also apply to the sorting_columns_ in the row groups?

rdblue replied:
I think sorting columns may need a separate update. These identifiers could potentially be used there, but I don't think the ordering (or direction) is currently stored.

rdblue commented:
I updated this in response to @julienledem's comments. I think he's right that we only need to store the order once for each column, so this now adds `list<ColumnOrder>` to the footer. The `Statistics` objects are unchanged because there's no need to add the order to each stats object in the format.

In the new commit, `ColumnOrder` is a struct 2 fields: the first field is `order`, which is an `Order` enum with `SIGNED`, `UNSIGNED`, and `CUSTOM` and the second field is `custom_order`, a string that describes the order used if `order` was `CUSTOM`.

rdblue commented:
I've updated the last commit to include the following:
* Added `ColumnOrder.is_ascending` boolean
* Added a specific format for the `custom_order` string

I had initially added `ICU54` to the `Order` enum, but it seems confusing to have `Order` include both specific orderings and classes of orderings, so I went with the string prefix option. We can change it if anyone is strongly opposed.

julienledem commented on src/main/thrift/parquet.thrift:
```diff
@@ -547,6 +547,61 @@ struct RowGroup {
+  0: required Order order;
+
+  /** Whether the order used is ascending (true) or descending (false) */
+  1: required boolean is_ascending;
```
SortingColumn already has this which defines the direction of the sort in the file:
``` 4: optional list<SortingColumn> sorting_columns```
In the case of statistics I don't think we should add an 'is_ascending' field in ColumnOrder as this basically just swaps min and max.

rdblue replied:
Works for me.

isnotinvain commented on src/main/thrift/parquet.thrift:
```diff
@@ -547,6 +547,61 @@ struct RowGroup {
+ */
+struct ColumnOrder {
+  /** The order used for this column */
+  0: required Order order;
```
in thrift we usually start field orderings at number 1 not 0. I don't remember why, but I see that we're doing that in the other structs so I think we should do it here too.

rdblue replied:
Fixed.

isnotinvain commented on src/main/thrift/parquet.thrift:
```diff
@@ -576,5 +631,8 @@ struct FileMetaData {
   6: optional string created_by
+
+  /** Sort order used for each column in this file */
+  7: optional list<ColumnOrder> column_orders;
```
How do you know from 1 ColumnOrder which column it applies to? It's position in this list has to match the position in the `schema` list above? If that's the case I think we should add a comment for that here. I guess this has to be here to avoid repeating it in each Statistics object?

rdblue replied:
There is one `ColumnOrder` for each column, in the same order as the column chunks (which is required to match the order in the schema).

julienledem replied:
Another option is to add an optional column_order field in SchemaElement which would set only for primitive types.
``` 10: optional ColumnOrder column_order; ```
https://github.com/rdblue/parquet-format/blob/9962df8e0ea85858cf032451d0ee83ec3f4d39fe/src/main/thrift/parquet.thrift#L260

rdblue replied:
My reasoning for not adding this to the schema is that I don't think I'd add this to the string-based schema definition. I want to keep those roughly the same by not adding things to SchemaElement that aren't in the schema strings.

julienledem replied:
good point

isnotinvain commented on src/main/thrift/parquet.thrift:
```diff
@@ -547,6 +547,61 @@ struct RowGroup {
+   * identify the actual order used for min, max, and soring values.
+   *
+   * This identifier should follow one of the following formats:
+   * * 'icu54:<locale-keyword>' - ICU 54 ordering for the ICU 54 locale keyword
```
what about something like:

```
struct ColumnOrder {
  1: required OrderType orderType
  2: optional string orderSubType
  3: required boolean is_ascending
}
```

```
enum OrderType {
  SIGNED, UNSIGNED, ICU_54, OTHER // more can be added as needed
}
```

and `orderSubType` isn't set for SIGNED/UNSIGNED ?

isnotinvain replied:
Actually, if we really want to avoid the confusion of mixing specific orders and classes of orders, we can use a union instead:

```
struct ColumnOrder {
  1: required Order order
  2: required boolean is_ascending
}

struct Signed {}
struct Unsigned {}
struct ICU_54 {
  1: required string locale_keyword
}
struct Other {
  1: required string name
}
union Order {
  1: Signed signed
  2: Unsigned unsigned
  3: ICU_54 icu_54
  4: Other
  // add more as needed
}
```

This will generate an Order struct that can only be one of `{Signed / Unsigned / ICU_54(locale_keyword) / Other(name)` and avoids having to deal with the fact that Signed / Unsigned require no additional info, and it also means we can add extra info to each type of ordering as needed.

rdblue replied:
I've removed the other symbols. We can revisit this when we know more about the requirements of processing engines.

julienledem replied:
I think we still need to finalize this aspect of the discussion.
the signed/unsigned aspect is what's blocking impala/sparksql right now.
When they move to implement collation it will be easier to converge on the collation string spec.
Let's open a separate JIRA for that.

isnotinvain replied:
In that case, should we remove the commented out fields + their docs?

julienledem replied:
Since adding an enum value is a breaking change in thrift and ordering is ignorable by old libraries that don't understand it, I'd favor a model like the one proposed by Alex above.
We need to verify that the union type implementation in thrift doesn't suffer from the same issue. A quick glance at the code seems to confirm it's fine but we just need to try it. 
In that case I think `1: required Order order` needs to be optional since it would be null when reading an unknown type.

If Union doesn't work, another approach (example bellow) is to simply have optional fields which we document as exclusive. However Union is cleaner.
```
struct ColumnOrder {
  // only one should be set
  1: optional boolean signed; // whether default signed unsigned binary comparison is used
  2: optional string sql_collation; // using sql collation string 
  3: optional string icu_54_locale; // using unicode locale
}
```

rdblue replied:
I would rather have a type that can be only one value, like a union or an enum. Otherwise there are more annoying cases that could cause problems: what would it mean when signed is true and sql_collation is set? It is better if we can guarantee only one, and I don't think that avoiding a single string field (rather than both sql_collation and icu_54_locale) is worth that ambiguity.

isnotinvain replied:
What did we decide? can we go with the union approach?

julienledem replied:
@rdblue It sounds like Union? +1 from me as well

rdblue replied:
I don't think we've confirmed that the union is forward-compatible. I haven't had time to check, so if either of you wants to that would be helpful.

julienledem replied:
Works for thrift 0.9.3. I don't have thrift 0.7 working on my laptop right now.
If you want to try with thrift 0.7:
https://github.com/apache/parquet-mr/pull/405

julienledem replied:
related: we should just update parquet-format to thrift 0.9.3
https://github.com/apache/parquet-format/pull/50

julienledem replied:
@rdblue: works for thrift 0.7 per travis-ci: https://github.com/apache/parquet-mr/pull/405

julienledem commented on src/main/thrift/parquet.thrift:
```diff
@@ -547,6 +547,58 @@ struct RowGroup {
+  /**
+   * Identifiers for custom orderings, to be defined in the ColumnOrder struct.
+   */
+  //CUSTOM = 2;
```
what would CUSTOM mean in that context?

lekv replied:
Impala and Hive store timestamps as a tuple (int64_t, uint32_t), where the first element is the time of the day in nanoseconds, and the second is the day according to the Julian calendar. Ordering these means to sort them by the second element first (the day), then by the first (the time). None of the other orderings seem to reflect this, so it would be good if we could at least set it to CUSTOM to signal that any validation will fail unless the client knows how to interpret them.

Does it make more sense to document something like IMPALA_TIMESTAMP as its own logical type?

xhochy replied:
We definitely should change the situation as currently the values stored as `INT96` aren't really interpretable as integers, better open a separate JIRA to untangle it a bit from the sorting discussion here. 

@lekv I also tried to document the current state here: https://github.com/apache/parquet-format/pull/49 would be nice if you also could have a short look if the formula is correct.

lekv replied:
@xhochy - I left some comments in #49. I agree that it's better to untangle / document Impala timestamps in a separate JIRA. Should we use PARQUET-861 for that, or shall I go ahead and make a new one?

@julienledem - Do you think it's better to defer adding "CUSTOM" orderings until adding the first LogicalType that needs one?

xhochy replied:
@lekv We can continue to do this in PARQUET-861, probably @rdblue wants also to comment in there. (knowing that he is in favor of deprecating INT96/timestamps as there are currently).

julienledem replied:
@lekv Yes, I'd rather defer adding CUSTOM then.

timarmstrong replied:
I don't understand the claim that INT96 aren't interpretable as integers. They don't map to a common hardware integer type but they're still logically a twos-complement integer.

I think on the Impala side we may end up implementing min-max stats for INT96 ahead of doing a larger rework of the timestamp data type, so it would be good to have clarity about what other implementations should do when encountered with INT96 min-max stats.

My take is that they should either ignore INT96 stats (if they don't want to implement a deprecated type) or use signed ordering. Parquet-mr doesn't do this currently. Does everyone agree that https://issues.apache.org/jira/browse/PARQUET-840 is a bug in parquet-mr?

timarmstrong replied:
I guess I'm assuming that signed INT96 ordering is equivalent to the logical IMPALA_TIMESTAMP ordering - I think this is true. because the least significant 8 bytes of the INT96 are the unsigned nanoseconds component and the most significant 4 bytes of the INT96 are the signed Julian day.

julienledem replied:
@timarmstrong if that's the case then we can just use SIGNED Ordering for them.
@lekv do you agree?

lekv replied:
@timarmstrong I assumed we are writing (nanoseconds, day) tuples, and the UNSIGNED order would not be equivalent to the logical ordering. Internally Impala stores them as [(time, date)](https://github.com/apache/incubator-impala/blob/master/be/src/runtime/timestamp-value.h#L255) and we don't reverse those. Hive seems to do it the same way as Impala: https://github.com/Parquet/parquet-mr/blob/fa8957d7939b59e8d391fa17000b34e865de015d/parquet-column/src/main/java/parquet/example/data/simple/NanoTime.java#L58

timarmstrong replied:
I think it works out because INT96 is little-endian - the most significant 4 bytes of the INT96 line up with the 4 bytes of the date

julienledem commented on LogicalTypes.md:
```diff
@@ -37,6 +37,8 @@ may require additional metadata fields, as well as rules for those fields.
 `UTF8` may only be used to annotate the binary primitive type and indicates
 that the byte array should be interpreted as a UTF-8 encoded character string.
 
+The sort order used for `UTF8` strings must be `UNSIGNED` byte-wise comparison.
```
+1

isnotinvain replied:
Is this true? I thought the sort order for strings *should* be unsigned, from now on, but it's valid and possible for old files to have signed ordering. Should we clarify that both orderings are supported but one is preferred?

rdblue replied:
This is stating that all UTF8 from now on should be written with UNSIGNED comparison. It is still the case that files without the SortOrder list used SIGNED.

julienledem commented on LogicalTypes.md:
```diff
@@ -55,6 +57,8 @@ allows.
 implied by the `int32` and `int64` primitive types if no other annotation is
 present and should be considered optional.
 
+The sort order used for signed integer types must be `SIGNED`.
```
My interpretation is we should fail. 
This is redundant with logical type but more explicit and deals with backward compatibility.

timarmstrong replied:
Sounds reasonable to me.

rdblue replied:
Any implementation should ignore stats that it cannot use. I think Lars mentioned in the recent hangout that there are transforms that can be done in some cases to use stats that are written with some compatible orderings.

A simple example is where a UINT32 field has min=10 and max=1000. Since both are positive and not in the top half of the UINT32 range, we know that the same min/max are produced by both signed and unsigned comparison and the implementation can use it.

julienledem commented on src/main/thrift/parquet.thrift:
```diff
@@ -576,5 +628,8 @@ struct FileMetaData {
   6: optional string created_by
+
+  /** Sort order used for each column in this file */
+  7: optional list<ColumnOrder> column_orders;
```
Yes, if column_orders is missing that means they are written the old way.
In the current PR it looks like old versions of the library won't tell the difference between old and new stats. If we want to make sure they don't see the new stats, we need to put them in a different field in the Stats struct. 
Do we want to do that?

timarmstrong replied:
+1 to making them a different field.

rdblue replied:
Sorry, I think this was clear from previous discussion, but not from the PR.

Readers must assume that the order is signed, unless this `SortOrder` is set because that is what the Java library currently writes. The original min and max stats fields are still used because the majority of the values are valid. There are only a few types that are currently stored incorrectly:
* UTF8 strings
* Decimals (stored as fixed or bytes)
* Intervals
* UINTs

Unsigned ints and intervals aren't used in any engine that I know about. Decimals are used, but I doubt anyone is filtering based on the stats because there isn't a way to pass a decimal to the filter code path in Parquet Java. (You'd have to pass an encoded decimal Binary; same problem for intervals.)

That means that the only type with incorrect stats that is really a problem is UTF8. The reason why we didn't catch this problem sooner is because characters stored as a single byte without the msb set (including ASCII) have the same sort order using signed and unsigned comparison. This is why Parquet Java has a property to ignore the wrong sort order and use stats anyway.

Using a separate field for the new stats would mean that old readers can't use stats for UTF8 fields, even though there are lots of cases where it is fine, and it is not really worse than what they do already -- behavior no one noticed was wrong for a year. It would also require more format changes and could prevent older readers from using stats for signed fields that are correct.

julienledem commented:
thanks for the review @timarmstrong.

isnotinvain commented on LogicalTypes.md:
```diff
@@ -98,6 +104,15 @@ integer. A precision too large for the underlying type (see below) is an error.
 A `SchemaElement` with the `DECIMAL` `ConvertedType` must also have both
 `scale` and `precision` fields set, even if scale is 0 by default.
 
+The sort order used for `DECIMAL` values must be `SIGNED`. The order is
```
+1

rdblue commented on LogicalTypes.md:
```diff
@@ -98,6 +104,15 @@ integer. A precision too large for the underlying type (see below) is an error.
+
+If the column uses `int32` or `int64` physical types, then signed comparison of
+the integer values produces the correct ordering. If the physical type is
+fixed, then the correct ordering can be produced by flipping the
```
Yes. Fixed is used interchangeably with the full name elsewhere as well.

timarmstrong replied:
Gotcha, I'm not totally familiar with some of these conventions.

rdblue commented on LogicalTypes.md:
```diff
@@ -98,6 +104,15 @@ integer. A precision too large for the underlying type (see below) is an error.
+the integer values produces the correct ordering. If the physical type is
+fixed, then the correct ordering can be produced by flipping the
+most-significant bit in the first byte and then using unsigned byte-wise
+comparison.
```
There is an optimization, but it is too long to describe here. The main point is that the order must match the signed order produced by comparing the decimal values. This paragraph only gives optimizations for easy types.

timarmstrong replied:
Makes sense, thanks for explaining.

rdblue commented on src/main/thrift/parquet.thrift:
```diff
@@ -547,6 +547,58 @@ struct RowGroup {
+/** Descriptor for the order used for min, max, and sorting values in a column
+ */
+struct ColumnOrder {
+  /** The order used for this column */
```
I don't think so. I think min and max are clear enough.

rdblue commented on LogicalTypes.md:
```diff
@@ -144,6 +169,9 @@ example, there is no requirement that a large number of days should be
+The sort order used for `INTERVAL` is `UNSIGNED`, produced by sorting by
+the value of months, then days, then milliseconds with unsigned comparison.
+
 ## Embedded Types
```
In the latest commit, I clarified that there isn't a required ordering for embedded types.

timarmstrong commented:
I agree it would be useful if the stats were usable by older
implementations when they're compatible.

Is there any way that a new implementation can write out new-style stats
and force older implementations to ignore them? E.g. if I want to use some
special unicode collation in my query engine, but also want to read the
files via an older parquet-mr, which assumes all stats are the old style.

On 15 February 2017 at 02:47, Ryan Blue <notifications@github.com> wrote:

> *@rdblue* commented on this pull request.
> ------------------------------
>
> In src/main/thrift/parquet.thrift
> <https://github.com/apache/parquet-format/pull/46>:
>
> > @@ -576,5 +628,8 @@ struct FileMetaData {
>     * e.g. impala version 1.0 (build 6cf94d29b2b7115df4de2c06e2ab4326d721eb55)
>     **/
>    6: optional string created_by
> +
> +  /** Sort order used for each column in this file */
> +  7: optional list<ColumnOrder> column_orders;
>
> Sorry, I think this was clear from previous discussion, but not from the
> PR.
>
> Readers must assume that the order is signed, unless this SortOrder is
> set because that is what the Java library currently writes. The original
> min and max stats fields are still used because the majority of the values
> are valid. There are only a few types that are currently stored incorrectly:
>
>    - UTF8 strings
>    - Decimals (stored as fixed or bytes)
>    - Intervals
>    - UINTs
>
> Unsigned ints and intervals aren't used in any engine that I know about.
> Decimals are used, but I doubt anyone is filtering based on the stats
> because there isn't a way to pass a decimal to the filter code path in
> Parquet Java. (You'd have to pass an encoded decimal Binary; same problem
> for intervals.)
>
> That means that the only type with incorrect stats that is really a
> problem is UTF8. The reason why we didn't catch this problem sooner is
> because characters stored as a single byte without the msb set (including
> ASCII) have the same sort order using signed and unsigned comparison. This
> is why Parquet Java has a property to ignore the wrong sort order and use
> stats anyway.
>
> Using a separate field for the new stats would mean that old readers can't
> use stats for UTF8 fields, even though there are lots of cases where it is
> fine, and it is not really worse than what they do already -- behavior no
> one noticed was wrong for a year. It would also require more format changes
> and could prevent older readers from using stats for signed fields that are
> correct.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/parquet-format/pull/46>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABGjqzo1Ku0yEvt3IH-NjbbKkg-ZKnGBks5rcdqQgaJpZM4Kwzz5>
> .
>

rdblue commented:
@timarmstrong, even after this PR, there's no way to record that you're using a special unicode collation, so it isn't allowed. If we included the custom order identifier, then readers that know to only use stats when they support the order are fine because they can ignore them. Older readers would not produce correct results.

This raises a good point about format compatibility. I think it is okay to continue using the current min and max fields because the result is no more broken than it is already. But, other collating sequences are forward-incompatible changes to the format if they use the same fields. Also, the proposals to add symbols to the `Order` struct are forward-incompatible changes: A reader that expects only `SIGNED` and `UNSIGNED` can't deserialize footers with new symbols.

That's enough to convince me that you guys were right all along and we should use new min and max fields. We can also store min and max in the old fields for older readers when compatible, but new stats that may have a different ordering shouldn't be available to older readers.

I'll add a commit that does the following:
* Adds new min and max fields to Statistics and notes that the old fields are optional and can be set when the sort order of the column is SIGNED.
* Add back the CUSTOM symbol to the Order enum and the custom_order string in ColumnOrder. Since we can't add symbols in the future, I think this is a good way to allow extension without forward-incompatible changes.

@isnotinvain and @julienledem, you'll probably want to take another look after the new commit.

julienledem commented:
adding new fields is fine by me. although we could rename min and max to min_signed and max_signed for clarity. Thrift allows renaming fields.
I put a comment above regarding the enum.

rdblue commented:
@julienledem, I updated this to use a union. One last review?

isnotinvain commented:
Union changes LGTM, thanks! +1

rdblue commented:
Thanks for the quick review, @isnotinvain! I'll merge early next week after everyone has had a chance to weigh in on the changes.

lekv commented:
Are we ready to merge this one?

rdblue commented:
I think it's ready, but I thought @julienledem may want to have another look.

lekv commented:
I just tried out the parquet.thrift file by copying it into the Impala sources and trying to compile and gcc choked on ConvertedType.NULL.

```
In file included from be/src/util/debug-util.h:34:0,
                 from be/src/catalog/catalog-util.cc:24:
be/generated-sources/gen-cpp/parquet_types.h:58:5: error: expected identifier before ‘__null’
     NULL = 25
     ^
be/generated-sources/gen-cpp/parquet_types.h:58:5: error: expected ‘}’ before ‘__null’
be/generated-sources/gen-cpp/parquet_types.h:58:5: error: expected unqualified-id before ‘__null’
In file included from be/src/util/debug-util.h:34:0,
                 from be/src/catalog/catalog-util.cc:24:
be/generated-sources/gen-cpp/parquet_types.h:236:3: error: ‘Type’ does not name a type
   Type::type type;
   ^
be/generated-sources/gen-cpp/parquet_types.h:241:3: error: ‘ConvertedType’ does not name a type
   ConvertedType::type converted_type;
   ^
```

Has anyone else tried to compile this with g++? We're currently using gcc-4.9.2 and thrift-0.9.0.

I'll investigate further but I also thought someone may have seen this before.

rdblue commented:
@lekv, ConvertedType.NULL is new, but was removed in #51 to avoid this problem. Reviews on that one are welcome.

lekv commented:
@rdblue - I missed it wasn't actually introduced in this commit. Apologies for the noise.

I also noticed that the current layout of the comments in `Statistics` seem to confuse the thrift compiler when turning them into a Python class comment.

```
class Statistics:
  """
  Statistics per row group and per page
  All fields are optional.

  Attributes:
   - max: DEPRECATED: min and max value of the column. Use min_value and max_value.

  Values are encoded using PLAIN encoding, except that variable-length byte
  arrays do not include a length prefix.

  These fields encode min and max values determined by SIGNED comparison
  only. New files should use the correct order for a column's logical type
  and store the values in the min_value and max_value fields.

  To support older readers, these may be set when the column order is
  SIGNED.
   - min
   - null_count: count of null value in the column
   - distinct_count: count of distinct values occurring
   - max_value: Min and max values for the column, determined by its ColumnOrder.

  Values are encoded using PLAIN encoding, except that variable-length byte
  arrays do not include a length prefix.
   - min_value
  """
```

Do we usually try to format the Thrift code in a way that the comments look nice in Python?

julienledem commented on src/main/thrift/parquet.thrift:
```diff
@@ -576,5 +607,14 @@ struct FileMetaData {
+   * be Signed. In addition, min and max values for INTERVAL or DECIMAL stored
+   * as fixed or bytes should be ignored.
+   */
+  7: optional list<ColumnOrder> column_orders;
```
Do you want to add a note that there is exactly one ColumnOrder per column as defined by the schema (1 column per leaf node in the schema)?

rdblue commented:
@julienledem, update this. One more review?

rdblue commented on src/main/thrift/parquet.thrift:
```diff
@@ -576,5 +607,14 @@ struct FileMetaData {
+   * Sort order used for each column in this file.
+   *
+   * If this list is not present, then the order for each column is assumed to
+   * be Signed. In addition, min and max values for INTERVAL or DECIMAL stored
```
If the order isn't present, then it was byte-wise signed comparsion that is useless. Readers should ignore the values.

rdblue commented:
I'll give others a day or two to weigh in and then commit this.

rdblue commented:
Merged. Thanks for all of the discussion and reviews, everyone!
